sw/device/tests/dif/dif_pinmux_test.cc

```diff
@@ -0,0 +1,246 @@
+    EXPECT_EQ(outsel_registers_.size(), PINMUX_MIO_OUTSEL_MULTIREG_COUNT);
+  }
+
+  void ExpectInitReset() {
+    EXPECT_READ32(PINMUX_REGEN_REG_OFFSET, 1);
+
+    for (const auto &reg : insel_registers_) {
+      EXPECT_WRITE32(reg, 0);
+    }
+
+    for (const auto &reg : outsel_registers_) {
+      EXPECT_WRITE32(reg, 0x2082082);
+    }
+  }
+
+  std::vector<ptrdiff_t> insel_registers_{
+      PINMUX_PERIPH_INSEL0_REG_OFFSET, PINMUX_PERIPH_INSEL1_REG_OFFSET,
+      PINMUX_PERIPH_INSEL2_REG_OFFSET, PINMUX_PERIPH_INSEL3_REG_OFFSET,
+      PINMUX_PERIPH_INSEL4_REG_OFFSET, PINMUX_PERIPH_INSEL5_REG_OFFSET,
+      PINMUX_PERIPH_INSEL6_REG_OFFSET,
+  };
+
+  std::vector<ptrdiff_t> outsel_registers_{
+      PINMUX_MIO_OUTSEL0_REG_OFFSET, PINMUX_MIO_OUTSEL1_REG_OFFSET,
+      PINMUX_MIO_OUTSEL2_REG_OFFSET, PINMUX_MIO_OUTSEL3_REG_OFFSET,
+      PINMUX_MIO_OUTSEL4_REG_OFFSET, PINMUX_MIO_OUTSEL5_REG_OFFSET,
+      PINMUX_MIO_OUTSEL6_REG_OFFSET,
+  };
+};
+
+TEST_F(InitTest, NullArgs) {
+  EXPECT_EQ(dif_pinmux_init(base_addr_, nullptr), kDifPinmuxBadArg);
+}
+
+TEST_F(InitTest, RegistersLocked) {
+  EXPECT_READ32(PINMUX_REGEN_REG_OFFSET, 0);
+
+  EXPECT_EQ(dif_pinmux_init(base_addr_, &dif_pinmux_), kDifPinmuxInitLocked);
+}
+
+TEST_F(InitTest, Success) {
+  ExpectInitReset();
+
+  EXPECT_EQ(dif_pinmux_init(base_addr_, &dif_pinmux_), kDifPinmuxInitOk);
+}
+
+class RegistersLockTest : public PinmuxTest {};
+
+TEST_F(RegistersLockTest, NullArgs) {
+  EXPECT_EQ(dif_pinmux_registers_lock(nullptr), kDifPinmuxBadArg);
+}
+
+TEST_F(RegistersLockTest, Success) {
+  EXPECT_WRITE32(PINMUX_REGEN_REG_OFFSET, 0);
+
+  EXPECT_EQ(dif_pinmux_registers_lock(&dif_pinmux_), kDifPinmuxOk);
+}
+
+class SetTests : public PinmuxTest {
+ protected:
+  // Tests should not rely on DIF defined constants and defines for setting
+  // expectations.
+  static constexpr uint32_t kExpectConstantZero = 0;
+
+  static constexpr uint32_t kExpectInselFirstIn = 2;
+  static constexpr uint32_t kExpectInselLastIn =
+      kExpectInselFirstIn + (PINMUX_PARAM_NMIOPADS - 1);
+
+  static constexpr uint32_t kExpectOutselFirstOut = 3;
+  static constexpr uint32_t kExpectOutselLastOut =
+      kExpectOutselFirstOut + (PINMUX_PARAM_NMIOPERIPHOUT - 1);
+
+  struct Register {
+    ptrdiff_t offset;    // Register offset from the base.
+    uint8_t last_field;  // Last field offset in the register.
+  };
+
+  // Set multireg expectations, one field per call for every multireg register.
+  void ExpectSetTests(const std::vector<Register> &regs, uint32_t field_width,
+                      uint32_t field_mask, uint32_t value) {
+    for (const auto &reg : regs) {
+      for (uint32_t i = 0; i <= reg.last_field; i += field_width) {
+        EXPECT_MASK32(reg.offset, {{i, field_mask, value}});
+      }
+    }
+  }
+};
+
+class InselSet : public SetTests {
+ protected:
+  InselSet() {
+    EXPECT_EQ(registers_.size(), PINMUX_PERIPH_INSEL_MULTIREG_COUNT);
+  }
+
+  std::vector<Register> registers_ = {
```

Comment:
You use `std::vector<T> foo{}` above but `std::vector<T> foo = {}`. Prefer the latter, because it avoids the other `std::vector<T>` ctors.